clients/da-vinci-client/src/main/java/com/linkedin/davinci/validation/PartitionTracker.java

```diff
@@ -19,136 +20,178 @@
   /**
-   * A logging throttler singleton for ProducerTracker with a 64KB bitset.
-   *
    * If an exception will be tolerated, there is no need to print a log for each single message;
    * we can log only once a minute. The error message identifier pattern for log throttling is:
    * topicName-partitionNum-exceptionType
    */
   private static final RedundantExceptionFilter REDUNDANT_LOGGING_FILTER =
-      RedundantExceptionFilter.getRedundantExceptionFilter(64 * 1024, DEFAULT_NO_REDUNDANT_EXCEPTION_DURATION_MS);
-
-  private final Logger logger;
+      RedundantExceptionFilter.getRedundantExceptionFilter();
 
-  private final GUID producerGUID;
-  // This will allow to create segments for different partitions in parallel.
-  protected final ConcurrentMap<Integer, Segment> segments = new VeniceConcurrentHashMap<>();
-  protected final ConcurrentMap<Integer, ReentrantLock> partitionLocks = new VeniceConcurrentHashMap<>();
+  private static final Function<GUID, ReentrantLockWithRef<Segment>> segmentProvider =
+      guid -> new ReentrantLockWithRef<>();
 
+  private final Logger logger;
   private final String topicName;
+  private final int partition;
+  private final Map<GUID, ReentrantLockWithRef<Segment>> segments = new VeniceConcurrentHashMap<>();
 
-  public ProducerTracker(GUID producerGUID, String topicName) {
-    this.producerGUID = producerGUID;
+  public PartitionTracker(String topicName, int partition) {
     this.topicName = topicName;
+    this.partition = partition;
     this.logger = LogManager.getLogger(this.toString());
   }
 
+  public int getPartition() {
+    return partition;
+  }
+
   public final String toString() {
-    return ProducerTracker.class.getSimpleName() + "(GUID: " + ByteUtils.toHexString(producerGUID.bytes()) + ", topic: "
-        + topicName + ")";
+    return PartitionTracker.class.getSimpleName() + "(topic: " + topicName + ", partition: " + partition + ")";
   }
 
-  public ReentrantLock getPartitionLock(int partition) {
-    return partitionLocks.computeIfAbsent(partition, key -> new ReentrantLock());
+  /** N.B. Intended for tests */
+  Set<GUID> getTrackedGUIDs() {
+    return Collections.unmodifiableSet(this.segments.keySet());
   }
 
   /**
-   * In some cases, such as when resetting offsets or unsubscribing from a partition,
-   * the {@link ProducerTracker} should forget about the state that it accumulated
-   * for a given partition.
-   *
-   * @param partition to clear state for
+   * @param guid for which to retrieve the lock and segment
+   * @return a non-null {@link ReentrantLockWithRef} (but which can have {@link ReentrantLockWithRef#containsRef()} == false)
    */
-  public void clearPartition(int partition) {
-    ReentrantLock partitionLock = getPartitionLock(partition);
-    partitionLock.lock();
-    try {
-      segments.remove(partition);
-    } finally {
-      partitionLock.unlock();
-    }
+  private ReentrantLockWithRef<Segment> getLockAndSegment(GUID guid) {
```

Comment:
This potentially doesn't require any more.
The original reason to introduce partition lock is for AA with aggregate mode to avoid race conditions for the same records from the same GUID.
Now, such mode doesn't exist any more, this locking logic can be cleaned up.

Today, records produced by the same GUID shouldn't appear in multiple RT topics.
But this optimization can be done later.

Reply:
Hmm... it'd be great to get rid of this locking as it would simplify the code even more. That being said, the previous structure should not have concurrent interactions on the segments either, since the messages coming from different RT topics would have affected different `ProducerTracker` instances, even if they hit the same partition within each `PT`… I wonder if we’re missing something about why this locking was there in the first place… :thinking:

Reply:
Talked offline. It does look like the locking isn't necessary anymore. It was there for a temporary phase of the AA migration, which is over now. We can remove it, but we'll do it in a future PR.